packages/blocks/src/store/process-block-type.js

```diff
@@ -0,0 +1,159 @@
+/**
+ * External dependencies
+ */
+import { isPlainObject } from 'is-plain-object';
+
+/**
+ * WordPress dependencies
+ */
+import deprecated from '@wordpress/deprecated';
+import { applyFilters } from '@wordpress/hooks';
+
+/**
+ * Internal dependencies
+ */
+import { isValidIcon, normalizeIconObject, omit } from '../api/utils';
+import { BLOCK_ICON_DEFAULT, DEPRECATED_ENTRY_KEYS } from '../api/constants';
+
+/** @typedef {import('../api/registration').WPBlockType} WPBlockType */
+
+const { error, warn } = window.console;
+
+/**
+ * Mapping of legacy category slugs to their latest normal values, used to
+ * accommodate updates of the default set of block categories.
+ *
+ * @type {Record<string,string>}
+ */
+const LEGACY_CATEGORY_MAPPING = {
+	common: 'text',
+	formatting: 'text',
+	layout: 'design',
+};
+
+/**
+ * Takes the unprocessed block type settings, merges them with block type metadata
+ * and applies all the existing filters for the registered block type.
+ * Next, it validates all the settings and performs additional processing to the block type definition.
+ *
+ * @param {string}      name          Block name.
+ * @param {WPBlockType} blockSettings Unprocessed block type settings.
+ *
+ * @return {WPBlockType | undefined} The block, if it has been processed and can be registered; otherwise `undefined`.
+ */
+export const processBlockType =
+	( name, blockSettings ) =>
+	( { select } ) => {
+		const blockType = {
+			name,
+			icon: BLOCK_ICON_DEFAULT,
+			keywords: [],
+			attributes: {},
+			providesContext: {},
+			usesContext: [],
+			selectors: {},
+			supports: {},
+			styles: [],
+			variations: [],
+			save: () => null,
+			...select.getBootstrappedBlockType( name ),
+			...blockSettings,
+		};
+
+		const settings = applyFilters(
+			'blocks.registerBlockType',
+			blockType,
```

Comment:
Filters might modify the `blockType` object directly, so a copy `{ ...blockType }` was passed in the past. If we aren't concerned with the modifications, then maybe it would be better to use `settings` instead of the `blockType` after this line.

Reply:
We don't need to make a copy because `blockType` is a new object (created from default, bootstrapped and unprocessed settings) on each call.

Before this PR, `processBlockType` was working directly on an object in Redux state, so a copy was needed.

Still, in both the old and new versions, there is a risk that a filter will do something like:
```js
blockType.supports.inserter = false;
```
mutating the unprocessed settings object in Redux. We'd need to make a deep copy to avoid that.

Reply:
That is a good point about deep copy. Thinking a bit more about all of these, I think the challenge here is that block deprecations have very specific needs. You essentially want to pass the orginal `blockType` without any modifications/mutations applied with  `blocks.registerBlockType` filter to avoid the situation where the deprecation has its opinions on how to handle the same filter for every deprecation.